azurerm/resource_arm_app_service_source_control_token.go

```diff
@@ -0,0 +1,119 @@
+package azurerm
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/Azure/azure-sdk-for-go/services/web/mgmt/2018-02-01/web"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/features"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/locks"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+var appServiceSourceControlTokenResourceName = "azurerm_app_service_source_control_token"
+
+func resourceArmAppServiceSourceControlToken() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmAppServiceSourceControlTokenCreateOrUpdate,
+		Read:   resourceArmAppServiceSourceControlTokenRead,
+		Update: resourceArmAppServiceSourceControlTokenCreateOrUpdate,
+		Delete: schema.Noop,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"BitBucket",
+					"Dropbox",
+					"GitHub",
+					"OneDrive",
+				}, false),
+			},
+
+			"token": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+		},
+	}
+}
+
+func resourceArmAppServiceSourceControlTokenCreateOrUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).web.BaseClient
+	ctx := meta.(*ArmClient).StopContext
+
+	log.Printf("[INFO] preparing arguments for App Service Source Control Token creation.")
+
+	scmType := d.Get("type").(string)
+	token := d.Get("token").(string)
+
+	locks.ByName(scmType, appServiceSourceControlTokenResourceName)
+	defer locks.UnlockByName(scmType, appServiceSourceControlTokenResourceName)
+
+	if features.ShouldResourcesBeImported() && d.IsNewResource() {
+		existing, err := client.GetSourceControl(ctx, scmType)
+		if err != nil {
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("Error checking for presence of existing App Service Source Control Token (Type %q): %s", scmType, err)
+			}
+		}
+
+		if existing.ID != nil && *existing.ID != "" {
+			return tf.ImportAsExistsError(appServiceSourceControlTokenResourceName, *existing.ID)
+		}
+	}
+
+	properties := web.SourceControl{
+		SourceControlProperties: &web.SourceControlProperties{
+			Token: utils.String(token),
```

Comment:
should we also be exposing the other fields here:

```
// SourceControlProperties sourceControl resource specific properties
type SourceControlProperties struct {
	// Token - OAuth access token.
	Token *string `json:"token,omitempty"`
	// TokenSecret - OAuth access token secret.
	TokenSecret *string `json:"tokenSecret,omitempty"`
	// RefreshToken - OAuth refresh token.
	RefreshToken *string `json:"refreshToken,omitempty"`
	// ExpirationTime - OAuth token expiration.
	ExpirationTime *date.Time `json:"expirationTime,omitempty"`
}
```

Reply:
Have checked the [Azure](https://docs.microsoft.com/en-us/azure/app-service/deploy-continuous-deployment#authorize-azure-app-service) and [Azure Stack](https://docs.microsoft.com/en-us/azure-stack/operator/azure-stack-app-service-configure-deployment-sources) documentation for configuring deployment sources, and made some more tests with the different source control providers. Based on this tests we should add the `token_secret` field. I cannot find any docs regarding `refresh_token` and `expiration_time`. I assume they should be set together, but don't know how to properly test it, since I don't know in what context they should be used. Let me know if we still should add `refresh_token` and `expiration_time`. Thanks!